### PR TITLE
SSL_add_chain_certificate_pem to add chain certificate on SslRef

### PR DESCRIPTION
**Pull request by iamwwc**
fix https://github.com/sfackler/rust-openssl/issues/1742

### CODE REVIEW

File: openssl/src/ssl/mod.rs [hunk 1]
```diff
@@ -3104,6 +3104,19 @@ impl SslRef {
             }
         }
     }
+    #[corresponds(SSL_add1_chain_cert)]
+    #[cfg(ossl102)]
+    pub fn add_chain_certificate_pem(&mut self, chain: &[u8]) -> Result<(), ErrorStack> {
+        let cert = X509::from_pem(chain)?;
+        let ret = unsafe {
+            ffi::SSL_add_chain_certificate_pem(self.as_ptr(), cert.as_ptr() as *mut _ as *mut _)
```
Comment (wiktor-k) [inline openssl/src/ssl/mod.rs:3112]: :thinking: in `as *mut _ as *mut _` isn't one `as *mut _` redundant?

Comment (iamwwc) [reply openssl/src/ssl/mod.rs]: thankyou, removed :)

File: openssl-sys/src/ssl.rs [hunk 2]
```diff
@@ -406,6 +408,10 @@ cfg_if! {
         }
     }
 }
+#[cfg(ossl102)]
+pub unsafe fn SSL_add_chain_certificate_pem(ssl: *mut ::SSL, ptr: *mut c_void) -> c_long {
```
Comment (sfackler) [inline openssl-sys/src/ssl.rs:412]: Does this exist as a macro in OpenSSL? Google does not return any results for `SSL_add_chain_certificate_pem`.

Comment (sfackler) [reply openssl-sys/src/ssl.rs:412]: It looks like it's called `SSL_add1_chain_cert` instead: https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/include/openssl/ssl.h.in#L1410

Comment (iamwwc) [reply openssl-sys/src/ssl.rs]: yes, it's a macro. I have rename it to reflect its macro name. but retain pem suffix

File: openssl/src/ssl/mod.rs [hunk 3]
```diff
@@ -3104,6 +3104,17 @@ impl SslRef {
             }
         }
     }
+    #[corresponds(SSL_add1_chain_cert)]
+    #[cfg(ossl102)]
+    pub fn add_chain_cert_pem(&mut self, chain: &[u8]) -> Result<(), ErrorStack> {
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs:3109]: `SSL_add1_chain_cert` doesn't take a byte buffer, it takes an `X509` object: https://www.openssl.org/docs/man1.1.1/man3/SSL_add1_chain_cert.html.

Please add a test.

Comment (iamwwc) [reply openssl/src/ssl/mod.rs]: done

**Comment by iamwwc** [comment]
sorry, I have rename another rust-openssl to tongsuo-rust-openssl to avoid open unused pull request on your repo accidental

this repo is full fork of your repo, so I reopen pr :)

**Comment by iamwwc** [comment]
@sfackler review?

**Pull request merged**